auth_mellon_util.c

```diff
@@ -1540,7 +1612,11 @@ bool am_validate_paos_header(request_rec *r, const char *header)
         goto cleanup;
     }
 
-    /* Split the header on the semicolon character */
+    /*
+     * Split the header on the semicolon character.
+     * There should be 2 items, the first is the version delclaration
+     * and the second is the service and it's options.
+     */
     semicolon_tokens = g_strsplit(header, ";", 0);
```

Comment:
You have a good point and I wondered about this myself. I did look at all the URN's used in SAML and none had commas or semicolons in them. As far as I know only URN's  will ever appear in this header. However the fact the URN's are quoted which according to the HTTP parsing rules turns them into a string token suggests the committee was forward looking to protect against embedded token separators inside a URN string token.

So I thought about the problem. Should I do real parsing, break the header into individual tokens and then operate on the sequence of tokens (maybe not as general as building a parse tree). My thought was, that is is awful complicated, especially just for one HTTP header. Writing robust parsing code in C is non-trivial (look I even made mistake in the  am_strip_whitespace_and_quotes function which you caught). I know we do a better job of parsing the headers than Shibboleth does because our code caught a mistake in both the SAML spec and in Shibbolth in the Accept header. The spec incorrectly used a semicolon separator when it should have been a comma and Shibboleth never caught because all they do is a string search for a substring and ignore everything else in the header. I've since gotten OASIS to fix the spec and filed a bug with Shibboleth.

My guess is this code is probably still more robust than Shibboleth, but you're right it's imperfect in the service of simplicity. I don't think we'll see URN's with embedded comma's or semicolons, none are currently defined, but of course there is no guarantee one won't be defined in the future.

I might take a crack at parsing the header into tokens and then doing a simple analysis of the grammar. This would also eliminate the need for the am_strip_whitespace_and_quotes() function which isn't 100% correct either because it accepts non-quoted strings and does not handle embedded quote characters which was another simplification I was willing to accept in the service of simplicity.